storage/spanner.go

```diff
@@ -0,0 +1,984 @@
+			if err != nil {
+				return err
+			}
+			if !status {
+				return errors.New("ConditionalCheckFailedException", tmpMap, expr)
+			}
+		}
+		tableConf, err := config.GetTableConf(table)
+		if err != nil {
+			return err
+		}
+		table = changeTableNameForSP(table)
+
+		pKey := tableConf.PartitionKey
+		pValue, ok := tmpMap[pKey]
+		if !ok {
+			return errors.New("ResourceNotFoundException", pKey)
+		}
+		var key spanner.Key
+		sKey := tableConf.SortKey
+		if sKey != "" {
+			sValue, ok := tmpMap[sKey]
+			if !ok {
+				return errors.New("ResourceNotFoundException", pKey)
+			}
+			key = spanner.Key{pValue, sValue}
+
+		} else {
+			key = spanner.Key{pValue}
+		}
+
+		mutation := spanner.Delete(table, key)
+		err = t.BufferWrite([]*spanner.Mutation{mutation})
+		if e := errors.AssignError(err); e != nil {
+			return e
+		}
+		return nil
+	})
+	return err
+}
+
+// SpannerBatchDelete - this delete the data in batch
+func (s Storage) SpannerBatchDelete(ctx context.Context, table string, keys []map[string]interface{}) error {
+	tableConf, err := config.GetTableConf(table)
+	if err != nil {
+		return err
+	}
+	table = changeTableNameForSP(table)
+
+	pKey := tableConf.PartitionKey
+	ms := make([]*spanner.Mutation, len(keys))
+	sKey := tableConf.SortKey
+	for i := 0; i < len(keys); i++ {
+		m := keys[i]
+		pValue, ok := m[pKey]
+		if !ok {
+			return errors.New("ResourceNotFoundException", pKey)
+		}
+		var key spanner.Key
+		if sKey != "" {
+			sValue, ok := m[sKey]
+			if !ok {
+				return errors.New("ResourceNotFoundException", sKey)
+			}
+			key = spanner.Key{pValue, sValue}
+
+		} else {
+			key = spanner.Key{pValue}
+		}
+		ms[i] = spanner.Delete(table, key)
+	}
+	_, err = s.getSpannerClient(table).Apply(ctx, ms)
+	if err != nil {
+		return errors.New("ResourceNotFoundException", err)
+	}
+	return nil
+}
+
+// SpannerAdd - Spanner Add functionality like update attribute
+func (s Storage) SpannerAdd(ctx context.Context, table string, m map[string]interface{}, eval *models.Eval, expr *models.UpdateExpressionCondition) (map[string]interface{}, error) {
+	tableConf, err := config.GetTableConf(table)
+	if err != nil {
+		return nil, err
+	}
+	colDLL, ok := models.TableDDL[changeTableNameForSP(table)]
+	if !ok {
+		return nil, errors.New("ResourceNotFoundException", table)
+	}
+	pKey := tableConf.PartitionKey
+	var pValue interface{}
+	var sValue interface{}
+	sKey := tableConf.SortKey
+
+	cols := []string{}
+	var key spanner.Key
+	var m1 = make(map[string]interface{})
+
+	for k, v := range m {
+		m1[k] = v
+		if k == pKey {
+			pValue = v
+			delete(m, k)
+			continue
+		}
+		if k == sKey {
+			delete(m, k)
+			sValue = v
+			continue
+		}
+		cols = append(cols, k)
+	}
+	if sValue != nil {
+		key = spanner.Key{pValue, sValue}
+	} else {
+		key = spanner.Key{pValue}
+	}
+	updatedObj := map[string]interface{}{}
+	_, err = s.getSpannerClient(table).ReadWriteTransaction(ctx, func(ctx context.Context, t *spanner.ReadWriteTransaction) error {
```

Comment:
See also below, if you were to use an `Update` mutation, you would not have to read the row first, and that again would mean that you could just call `Client.Apply` with the mutation instead of having to define a `ReadWriteTransaction`.

Reply:
@rosspatil your input?